[device_mesh] add back the private init backend option

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #124780

This PR adds back the private init backend option (we had `_init_process_groups` before), to tackle the issues sub
mesh creation. This is a regression fix to 2.3 as we removed the `_init_process_groups` option in 2.3, which triggers a lot more sub process group creations, potentially causing memory spikes

in device mesh slicing we don't want to create process groups again,
so explicitly turn the group creation off it's useful

Also I think there might be more submesh creation functionality so
having this flag would ensure that there's no new group created

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

Differential Revision: [D56497780](https://our.internmc.facebook.com/intern/diff/D56497780)

wanchaol committed:
[device_mesh] add a private init backend option

This PR adds a private init backend option, to tackle the issues sub
mesh creation:

in device mesh slicing we don't want to create process groups again,
so explicitly turn the group creation off it's useful

Also I think there might be more submesh creation functionality so
having this flag would ensure that there's no new group created

[ghstack-poisoned]

awgu commented:
Oof, this is helpful. I was just writing the same code locally 😅

awgu reviewed (approved):
SGTM!

awgu commented on torch/distributed/device_mesh.py:
```diff
@@ -222,9 +224,9 @@ def __init__(
+            # Skip process group initialization if xla device or init backend is False
             # TODO(yeounoh) implement DeviceMesh backend and register XLA backend.
-            if device_type != "xla":
+            if device_type != "xla" and _init_backend:
```
If we skip `_get_or_create_default_group()`, then `self._coordinate_on_dim` will not be defined when `_init_backend=False`.

wanchaol replied:
oh I think we should fix this it seems, I'll do it as a follow up PR

wz337 replied:
I think I did similar thing for nD submesh slicing and not initing the sub pgs again for the mesh slice won't work as there is dim_group_infos for the mesh slice. iirc, that's why I need to revert.

awgu replied:
Yes, upon more investigation, I think we need to populate both `self._coordinate_on_dim` and `self._dim_group_infos`, so this current change should break unit tests IIUC.

wz337 replied:
It won't break any test yet as _init_backend is default to True, but a followup is needed for most of the API to work with _init_backend = False.

wanchaol replied:
I think we would always want the default `_init_backend` to be True, only when we do submesh slicing we would manually turn it off

awgu commented on torch/distributed/device_mesh.py:
```diff
@@ -89,6 +89,7 @@ def create_child_mesh(
                     device_mesh.device_type,
                     mesh_1d,
                     mesh_dim_names=(mesh_dim_name,),
+                    _init_backend=False,
```
@wz337 Even though `_init_backend` is defaulted to `True`, here we are passing it as `False` for child meshes. For FSDP2 2D training, I was thinking that this means the child meshes are now missing the `self._coordinate_on_dim` and `self._dim_group_infos`.

wanchaol replied:
there seems to have some test failures, so trying to fix it directly in this PR

wanchaol replied:
tested locally should fixed those tests

awgu replied:
How about `_self._dim_group_infos`? I think we need that to support `.get_group()`.

wanchaol replied:
those are already populated when we create the submesh?https://github.com/pytorch/pytorch/blob/main/torch/distributed/device_mesh.py#L96

awgu replied:
Ah, got it! Makes sense!

wanchaol replied:
basically when use `init_backend=False` I think there would be some manual pg creation or pg reuse, so we would need to explicitly attach the dim_group_info after `DeviceMesh` constructor?

i.e. for things like `from_group`, inside this API after `DeviceMesh` constructed, we need to attach the dim_group_info to the DeviceMesh created, the `dim_group_info` need to be generated in a similar way as https://github.com/pytorch/pytorch/blob/main/torch/distributed/device_mesh.py#L314-L319

wanchaol committed:
Update on "[device_mesh] add a private init backend option"


This PR adds a private init backend option, to tackle the issues sub
mesh creation:

in device mesh slicing we don't want to create process groups again,
so explicitly turn the group creation off it's useful

Also I think there might be more submesh creation functionality so
having this flag would ensure that there's no new group created

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

wanchaol committed:
Update on "[device_mesh] add a private init backend option"


This PR adds a private init backend option, to tackle the issues sub
mesh creation:

in device mesh slicing we don't want to create process groups again,
so explicitly turn the group creation off it's useful

Also I think there might be more submesh creation functionality so
having this flag would ensure that there's no new group created

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

wz337 commented:
@wz337 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D56497780).

wanchaol commented:
@pytorchbot merge

mvpatel2000 commented:
@wanchaol can we include this in torch 2.3? https://github.com/pytorch/pytorch/issues/125425 it would be nice to fix this regression

wanchaol commented:
> @wanchaol can we include this in torch 2.3? #125425 it would be nice to fix this regression

@mvpatel2000 Yes thanks for reminding! I'll submit a cherry pick soon
